Add funccount support to perf_test

iomartin opened the pull request:
This will count how many times the fast/slow path were taken. The slow path counter is an approximation.

This will require a [patched funccount](https://github.com/brendangregg/perf-tools/pull/115) to work.

iomartin committed:
perf_test: cleanup adding cpu results

Remove code duplication

iomartin committed:
perf_test: convert runtime,ramptime to seconds

fio supports different time suffixes.

Convert all to seconds (without an unit) to make it easier for funccount
to be able to measure for the right amount of time (in a future patch).

iomartin committed:
funccount: introduce funccount wrapper

funccount is a script from perf-tools[1] that uses ftrace to count how
many times kernel functions were called.

This python module is a tiny wrapper that makes it convenient to use and
returns a dictionary of all functions requested and their counts.

Note that in order to trace multiple functions, a fork of funccount is
required[2].

The optional sleep parameter will be useful when this is used with the
perf_tool, so that we can ignore IO done during the fio ramp_time.

[1]: https://github.com/brendangregg/perf-tools
[2]: https://github.com/brendangregg/perf-tools/pull/115

iomartin committed:
perf_test: add funccount support

Add a --funccount option to count how many IOs go through the slow/fast
paths.

Note that the slow path is only an approximation, as there isn't a
function that is always (and only) called by the slow path.

This is only available for the write path, as all functions that we
could potentially use for the fast read path were inlined.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,30 @@
+        self.sleep = sleep
+
+    def start(self):
+        cmd = f"sleep {self.sleep} && funccount {self.funcs}"
```
If you're going to use shell=True, then you need to be careful about quoting your arguments, as you would with a shell script....

But I try to avoid using the shell inside a python script if at all possible. A thread with a delay on a stop Event before calling funccount would probably be a bit nicer.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,30 @@
+import subprocess as sp
+import time
+
+class FUNCCOUNT:
```
Why is this class all caps?

iomartin replied:
I was following the fio.py but I renamed to capitalize just the first letter

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,30 @@
+
+    def stop(self):
+        pgid = os.getpgid(self.run.pid)
+        os.killpg(pgid, signal.SIGINT)
```
Do we need this nasty process group stuff? Possibly if it's not in a shell we can just use self.proc.send_signal()...

self.run seems like a poor choice for a variable name

iomartin replied:
I couldn't get it to work without the process group stuff. I also tried doing this outside of Python but sending a `kill -INT <pid>` didn't work. strace shows this:

```
0x7ffdebf1be20, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_USER, si_pid=93677, si_uid=0} ---
rt_sigreturn({mask=[CHLD]})             = -1 EINTR (Interrupted system call)
```

lsgunth replied:
Ah. The funccount script traps on interrupt and only sets quit=1 when that happens. Then it continues until the sleep command finishes.

A ctrl-c usually sends SIGINT to all processes in the foreground on the terminal which would include the sleep command.

So yes this seems necessary. But we should put a comment to that effect.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,30 @@
+            time.sleep(0.1)
+
+        res = {}
+        for line in self.run.stdout.readlines():
```
Really seems like we should be using Popen.communicate() here....

If funccount outputs more data than what is buffered this will hang in the while loop above (which is a poor substitute for Popen.wait())....

lsgunth reviewed (commented):
What happens if funccount is not installed? It looks like we will just get an exception instead of a useful message.

Probably same problem with fio (though that's going to be more commonly installed).

iomartin commented:
Fixed all the suggestions

iomartin commented:
I need to handle invalid functions to be traced better, please wait for that to review (probably only tomorrow)

iomartin commented:
I added a check during the __init__ to see if the request functions are available.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,54 @@
+        while self.proc.poll() is None:
+            time.sleep(0.1)
+
+        out, err = self.proc.communicate()
```
This is still broken in the same way. If the program stalls because it can't write to stdout (because you haven't read it yet) then the poll loop above will hang.

communicate() waits for the process to finish so you can just drop that loop.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,54 @@
+            time.sleep(0.1)
+
+        out, err = self.proc.communicate()
+        if err:
```
probably better to check self.proc.returncode than assume output to stderr is a failure.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,54 @@
+
+        out, err = self.proc.communicate()
+        if err:
+            raise RuntimeError(f'funccount failed:\n\n{err}')
```
RuntimeError? So if I want to catch this I have to somehow filter out other python cases that might throw this error?

How about subprocess.calledprocesserror() or some such so it matches what others might expect.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,54 @@
+            r = re.compile(func.replace('*', '.*'))
+            if any(r.match(line) for line in avail_funcs):
+                continue
+            raise ValueError(f'Function/Regex not available for tracing: {func}')
```
Similarly here. Don't reuse bultin exceptions or you might catch things you didn't expect. In this case I'd probably make a specific Exception type.

lsgunth commented on perf_test:
```diff
@@ -142,6 +162,24 @@ if __name__ == "__main__":
+            try:
+                fc = Funccount(f'raid5_make_request raid5_write_one_full_stripe',
+                               sleep=args.ramptime)
+            except ValueError as e:
```
Better to catch exceptions around all of the main code (which is getting quite large and could use factoring out into other functions).

You should also be catching subprocess errors and FileNotFoundErrors (which will indicate if funccount is not found or fio is not found) and handling them appropriately.

iomartin committed:
perf_test: catch exception when fio is not installed

iomartin commented:
Fixed everything.

I also changed this PR's target to be on top of #25 as this series can use the cleanup done on #25 and to simplify a future merge.

lsgunth commented on perf_test:
```diff
@@ -39,24 +39,27 @@ def run_fio(fio, bs, bw, write, offset_increment=0, rebuild=False,
+                                   "raw"    : job_res["iops"],
+            }
+    except FileNotFoundError:
+        sys.exit('Could not find fio.')
```
Again, put the try/finally over everything and just print the exception. Don't try to narrowly pick out exceptions and assume the file not found is fio.

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,84 @@
+
+    def __init__(self, funcs, executable='funccount', sleep=0):
+        if not shutil.which(executable):
+            raise FileNotFoundError(executable)
```
Subprocess functions will raise this error if the executable isn't found.... I'd rely on that.

But generrally raising an exception includes a more detailed message, not just the name of the executable.

I was going to suggest just calling subprocess.check_call([executable, "--help"]) ahead of time to check if the executable exists before running everything. But in this application, if funccount doesn't exist maybe it's best to just not include the results and print a warning at the end.

lsgunth commented on perf_test:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 
 from fio import FIO
+import funccount
```
I feel like we don't need to proliferate more files.... Why not just leave the funccount class in this file until there's a good reason for it to be reused somewhere?

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,84 @@
+    def _sleep_and_run(self):
+        self.event.wait(self.sleep)
+
+        self.lock.acquire()
```
The preferred syntax is `with self.lock`.... But this locking stuff seems awfully complicated.

And:

 * Probably shouldn't start the process if the "event" is terminated early
 * No need to lock or try to race, just stop the event and join the thread; creating the process is fast.
 * Should pass any exceptions from the thread to the main thread
 * I usually find it is better to subclass Thread than use a specific method like this.
 
So roughly something like:

```python
class FuncCount(threading.Thread):
    def __init__(self, funcs, sleep=0, executable="./funccount2"):
        super().__init__()
        self._cmd = [executable, funcs]
        self._sleep = sleep

        self._stop_event = threading.Event()
        self._proc = None
        self._excp = None

    def run(self):
        if self._stop_event.wait(self._sleep):
            return

        try:
            self._proc = subprocess.Popen(self._cmd, text=True,
                                          stdout=subprocess.PIPE,
                                          stderr=subprocess.PIPE,
                                          start_new_session=True)
        except Exception as e:
            self._excp = e

    def stop(self):
        self._stop_event.set()
        self.join()

        if self._excp:
            raise self._excp

        if not self._proc:
            return

        pgid = os.getpgid(self._proc.pid)
        os.killpg(pgid, signal.SIGINT)

        out, err = self._proc.communicate(timeout=1)
        #etc

```

lsgunth commented on funccount.py:
```diff
@@ -0,0 +1,84 @@
+                                 text=True, start_new_session=True)
+        self.lock.release()
+
+    def __init__(self, funcs, executable='funccount', sleep=0):
```
So, in debian, there are two funccounts with different names (that may behave slightly differently).

There's funccount-perf from the perf-tools-unstable package and funccount-bpfcc from the bpfcc-tools package.

There's no way to get a 'funccount' pure executable in Debian.

So I'd rather suggest, by default we search for all three possibilities and use whatever is installed.

lsgunth commented on perf_test:
```diff
@@ -152,6 +194,16 @@ if __name__ == "__main__":
+        print(f'Failed to setup funccount: {e}', file=sys.stderr)
+        fc = None
+    except FileNotFoundError as e:
+        print('Could not find funccount. Please install from https://github.com/brendangregg/perf-tools/pull/115',
```
Oh, yikes, we need the specific custom version from that pull request.... That is really not something we want as it creates a giant pain for anyone trying to use this script. If this was really necessary, I'd almost just make my own implementation of funccount in python here (it's pretty simple)... But I don't think that's necessary...

Seems like your change is not actually needed. I was a bit confused by the argument passing as this code is passing only a single argument with spaces (which eventually gets passed directly to set_ftrace_filter). Then your PR to funccount seems to just make it so it can accept multiple arguments and and pass them as one to set_trace_filter.... This doesn't seem like a requirement.


```
sudo funccount-perf 'vfs_write vfs_open'
```
Works as one would expect.

If we use the bpfcc version then we have to do:

```
sudo funccount-bpfcc  'vfs_write|vfs_open'
```

Because it actually uses a regular expresion to filter the available functions in the python script.



So let's just plan on using standard funccount tools... Not link to unecessary PRs or depend on unreleased software.
